include/utils.h

```diff
@@ -458,7 +526,22 @@ inline bool IsStructSequenceClassImpl(const py::handle& type) {
     return false;
 }
 inline bool IsStructSequenceClass(const py::handle& type) {
-    return PyType_Check(type.ptr()) && IsStructSequenceClassImpl(type);
+    if (PyType_Check(type.ptr())) [[likely]] {
+        static auto cache = std::unordered_map<py::handle, bool, TypeHash, TypeEq>{};
+        auto it = cache.find(type);
+        if (it != cache.end()) [[likely]] {
+            return it->second;
+        }
+        bool result = IsStructSequenceClassImpl(type);
+        cache.emplace(type, result);
+        (void)py::weakref(type, py::cpp_function([type](py::handle weakref) -> void {
+                              cache.erase(type);
+                              weakref.dec_ref();
+                          }))
+            .release();
+        return result;
+    }
+    return false;
 }
```

Comment:
Is this the best way to cache Python objects while cooperating with the Python GC? Could you have a look of this @Skylion007? Thanks!